### PR TITLE
fix(codegen): physicalize packed FP4 tensor addressing

### LINKED ISSUE #2754: [Bug] Packed FP4 ND tensors use logical width as the GM row stride on A5
## Background

In pypto-lib, a two-row packed FP4 cache tensor on A5 has a precision problem:
row 1 is addressed with a 512-byte step for 512 logical FP4 elements, while its
`torch.float4_e2m1fn_x2` carrier row contains 256 bytes.

Reproduce with:

    python tmp_fp4_stride_repro.py -p a5 -d 0

The reproduction case is not committed. Save the source below as
`tmp_fp4_stride_repro.py` in a public pypto-lib checkout.

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `c3f0dea274f55d9648920f17c968e8564ae9fcdc` (branch: `main`) |
| pypto | `ddf42145191aa8c8fcffc563d2ee644364c22f28` (branch: `main`) |
| simpler | `39ce891dbb3f665e72e99b4a1387b47012fb18bd` (detached) |
| ptoas | `0.61` |
| pto-isa | `5a4f74cbf627d4aac2e0ce10d5e0d8b118343265` (detached) |
| CANN | `9.2.0` |

All installed components match the versions pinned by the current PyPTO
checkout.

Diagnosis: **pypto Codegen**. The JIT boundary correctly expands the physical
x2 carrier shape to 512 logical FP4 elements, and the generated Vec tile
correctly has 256 carrier columns. However, tensor-view lowering emits a GM row
stride and row offset of 512 carrier elements.

### Component

Codegen

### Description

A contiguous ND tensor declared as `pl.Tensor[[2, 512], pl.FP4]` is backed by a
Torch tensor with physical shape `[2, 256]` and dtype
`torch.float4_e2m1fn_x2`. A simple per-row load/store copies row 0 correctly but
reads and writes row 1 at the wrong address.

This prevents packed FP4 cache tensors from using their production PyPTO
signatures. pypto-lib currently has to expose them as
`pl.UINT8[..., logical_width // 2]` and manually pack/unpack E2M1 nibbles.

Related: #2087.

### Steps to Reproduce

<details>
<summary>Reproduction case: <code>tmp_fp4_stride_repro.py</code></summary>

```python
import argparse

import pypto.language as pl
import torch

from golden import TensorSpec, run


ROWS = 2
LOGICAL_WIDTH = 512
PHYSICAL_WIDTH = LOGICAL_WIDTH // 2


@pl.jit
def fp4_copy(
    source: pl.Tensor[[ROWS, LOGICAL_WIDTH], pl.FP4],
    output: pl.Out[pl.Tensor[[ROWS, LOGICAL_WIDTH], pl.FP4]],
):
    for row in pl.spmd(ROWS):
        value = pl.load(source, [row, 0], [1, LOGICAL_WIDTH])
        pl.store(value, [row, 0], output)
    return output


def values():
    physical = torch.empty((ROWS, PHYSICAL_WIDTH), dtype=torch.uint8)
    physical[0] = torch.arange(PHYSICAL_WIDTH, dtype=torch.uint8)
    physical[1] = torch.arange(PHYSICAL_WIDTH, dtype=torch.uint8).bitwise_xor(0xA5)
    return physical.view(torch.float4_e2m1fn_x2)


def golden_copy(tensors):
    tensors["output"].view(torch.uint8).copy_(tensors["source"].view(torch.uint8))


def compare_bytes(actual, expected, **_kwargs):
    actual_bytes = actual.view(torch.uint8)
    expected_bytes = expected.view(torch.uint8)
    mismatches = torch.where(actual_bytes != expected_bytes)
    if not mismatches[0].numel():
        return True, ""
    first = tuple(index[0].item() for index in mismatches)
    return False, (
        f"first byte mismatch at {first}: "
        f"actual={actual_bytes[first]}, expected={expected_bytes[first]}"
    )


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", default="a5")
    parser.add_argument("-d", "--device", type=int, required=True)
    args = parser.parse_args()
    source = values()
    specs = [
        TensorSpec("source", list(source.shape), source.dtype, init_value=source),
        TensorSpec("output", list(source.shape), source.dtype),
    ]
    result = run(
        fp4_copy,
        specs,
        golden_fn=golden_copy,
        compare_fn={"output": compare_bytes},
        config=dict(platform=args.platform, device_id=args.device),
        rtol=0.0,
        atol=0.0,
    )
    if result.error:
        print(result.error)
    raise SystemExit(0 if result.passed else 1)
```

</details>

Run on a real A5 through the normal device allocation mechanism.

### Expected Behavior

Both physical rows are copied byte-for-byte. For logical width 512, advancing
one row must advance 256 `float4_e2m1fn_x2` carrier elements.

### Actual Behavior

The first mismatch is the first byte of row 1:

```text
first byte mismatch at (1, 0): actual=0, expected=165
```

Generated PTO/C++ uses logical width as the carrier stride:

```text
shape = [2, 512], strides = [512, 1]
Tile<..., float4_e2m1x2_t, 1, 512, ...> physical cols = 256
GlobalTensor<float4_e2m1x2_t, ..., Stride<512, 512, 512, 512, 1>>
row pointer offset = row * 512
```

### Git Commit ID

`ddf42145191aa8c8fcffc563d2ee644364c22f28`

### NPU Kind

Other: Ascend A5

### Host Platform

Linux (aarch64)

### Additional Context

Python `3.11.16`, Torch `2.14.0+cpu`.

A regression test should cover at least two rows of a contiguous ND FP4 tensor
and compare its underlying x2 carrier bytes. The fix must keep logical FP4
shapes in IR while converting non-last-axis strides and offsets to carrier
units before pointer arithmetic.

### PR DESCRIPTION
**Pull request by yanghaoran29**
## Summary

- keep packed FP4 transfer shapes and strides logical for PTO-ISA multi-row loads/stores
- physicalize each non-zero FP4 partition origin with a carrier-unit `pto.addptr`, then create a logical view and an all-zero partition
- reject packed-axis origins that are not statically known and even, and reject unprovable explicit dynamic outer strides
- cover parameter views, body-created `tensor.view`, the existing MXFP4 multi-row path, and the A5 byte-exact two-row reproduction

## Testing

- `cmake --build build --parallel 2`
- `python tests/lint/clang_tidy.py --strict-version --jobs=2 --diff-base=upstream/main`
- `pre-commit run --hook-stage commit`
- `python -m pytest tests/ut/codegen/test_pto_codegen.py tests/ut/codegen/test_mx_ops_codegen.py tests/ut/codegen/test_orchestration_codegen_more.py tests/ut/ir/transforms/test_init_memref.py -n 2 -q` (250 passed)
- `python -m pytest tests/st/runtime/ops/test_matmul_mx.py -v --platform a5 --execute-via-task-submit --task-submit-device 3` (2 passed; AICPU affinity limited to cores 1-3)
- `python -m pytest tests/st/runtime/ops/test_fp4_copy.py -v --platform a5 --execute-via-task-submit --task-submit-device 3` (1 passed; AICPU affinity limited to cores 1-3)

Closes #2754

### CODE REVIEW

File: src/backend/common/pto_ops_shared.cpp [hunk 1]
```diff
@@ -138,6 +138,35 @@ std::vector<std::string> GetIndexOffsetCodes(const std::vector<ir::ExprPtr>& exp
...
+    return codes;
+  }
+
+  // The generic helper above has already converted and clamped this value to
+  // a non-negative index.  Dynamic packed offsets are produced by affine tile
+  // loops in normal lowering; divide here at the last boundary before PTOAS
+  // pointer arithmetic.  Statically odd origins are rejected above.
+  const std::string two = codegen.GetOrEmitConstant(static_cast<int64_t>(2), DataType::INDEX);
+  const std::string carrier_offset = codegen.NewTemp();
+  codegen.Emit(carrier_offset + " = arith.divui " + codes.back() + ", " + two + " : index");
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply src/backend/common/pto_ops_shared.cpp]: Fixed — packed FP4 last-axis offsets now require a statically known even value, so a runtime-dependent odd origin cannot be silently rounded down. Added a regression test for the rejection.

File: tests/st/runtime/ops/test_fp4_copy.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    platform="a5",
+)
+# Keep this A5 regression on the machine's supported AICPU affinity set [1, 2, 3].
+_FP4_COPY_CASE.config.aicpu_thread_num = 3
+
+
+@pytest.mark.platforms("a5")
+@st.cases(_FP4_COPY_CASE)
+def test_fp4_copy_two_rows(case_run):
+    case_run.assert_passed()
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply tests/st/runtime/ops/test_fp4_copy.py:69]: Fixed — added the standard direct pytest entry point with `--platform a5`.

File: src/backend/common/pto_ops_shared.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply src/backend/common/pto_ops_shared.cpp]: Addressed. The revised lowering rejects every dynamic packed-axis FP4 origin before carrier conversion, so no odd runtime value can be truncated or alias another nibble. The regression test remains in place.

File: src/codegen/pto/pto_codegen.cpp [hunk 3]
```diff
@@ -1419,6 +1420,35 @@ void PTOCodegen::EmitMakeTensorViews(const FunctionPtr& func) {
...
+          stride_names[j] = GetOrEmitConstant(constant_stride->value_ / 2, DataType::INDEX);
+          continue;
+        }
+        std::string carrier_stride = NewNamedTemp(param->name_hint_ + "_carrier_s" + std::to_string(j));
+        Emit(carrier_stride + " = arith.divui " + stride_names[j] + ", " +
+             GetOrEmitConstant(static_cast<int64_t>(2), DataType::INDEX) + " : index");
+        stride_names[j] = std::move(carrier_stride);
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply src/codegen/pto/pto_codegen.cpp]: Addressed by removing this stride-conversion path. PTO-ISA TLOAD/TSTORE must receive logical FP4 strides; only non-zero partition base addresses are converted to carrier units now. Explicit dynamic outer strides are rejected when such an address shift is needed, while canonical dynamic ND strides remain provably even because they include the packed FP4 last extent.

**Comment by yanghaoran29** [comment]
Addressed the remaining CodeRabbit review summary, including the outside-diff `tensor.view` case. The revised design keeps logical strides for PTO-ISA transfers and shifts non-zero FP4 partition origins with a carrier-unit `pto.addptr`; this applies to both parameter views and body-created views. Added direct coverage for body-created `tensor.view` and for zero-origin multi-row FP4 transfer, then verified the original two-row copy and the existing MXFP4 matmul on A5 with 3 AICPU cores.

File: docs/en/dev/codegen/00-pto_codegen.md [hunk 4]
```diff
@@ -352,6 +352,19 @@ Example for a `[16, 1]` column vector (no DN annotation in DSL):
     : !pto.tensor_view<?x?xf32>
 ```
 
+#### Packed FP4 Carrier Addressing (A5)
```
[the remark this answers is not in the document]
Comment (yanghaoran29) [reply docs/en/dev/codegen/00-pto_codegen.md:355]: Fixed. Condensed the packed-FP4 section without dropping the split logical-stride/carrier-pointer semantics; the English guide is now exactly 1,000 lines. Documentation parity and markdown checks pass.
